File: problem_0071/Solution.java
/*
Spiral Order Matrix I

Given a matrix of m * n elements (m rows, n columns), return all elements of the matrix in spiral order.

Example:

Given the following matrix:

[
    [ 1, 2, 3 ],
    [ 4, 5, 6 ],
    [ 7, 8, 9 ]
]

You should return

[1, 2, 3, 6, 9, 8, 7, 4, 5]
*/

import java.io.*;
import java.util.*;

public class Solution {

    public ArrayList<Integer> my_spiralOrder(
        final List<ArrayList<Integer>> matrix)
    {
        ArrayList<Integer> res = new ArrayList<Integer>();

        int m = matrix.size();
        if (m == 0) return res;
        int n = matrix.get(0).size();
        if (n == 0) return res;

        for (int i = 0; i < m/2; ++i)
        {
            for (int j = i; j < n-i; ++j)
            {
                res.add(matrix.get(i).get(j));
            }

            for (int j = i+1; j < m-1-i && 2*i <= n-1; ++j)
            {
                res.add(matrix.get(j).get(n-1-i));
            }

            for (int j = n-1-i; j >= i; --j)
            {
                res.add(matrix.get(m-1-i).get(j));
            }

            for (int j = m-2-i; j > i && 2*i < n-1; --j)
            {
                res.add(matrix.get(j).get(i));
            }
        }

        if ((m & 1) != 0)
        {
            int i = m >> 1;
            for (int j = i; j < n-i; ++j)
            {
                res.add(matrix.get(i).get(j));
            }
        }

        return res;
    }

    public ArrayList<Integer> spiralOrder(final List<ArrayList<Integer>> A) {
        ArrayList<Integer> result = new ArrayList<Integer>();

	int m, n;
	m = A.size();
	n = A.get(0).size();

        if (m == 0)
	    return result;
	int len;
	int dir = 0; // right
	int row, col, layer;
	row = col = layer = 0;
	result.add(A.get(0).get(0));

        for (int step = 1; step < m * n; step++) {
            switch(dir) {
                // Go right
                case 0:
                    if (col == n - layer - 1) {
                        dir = 1;
                        row++;
                    } else {
                        col++;
                    }
                    break;
                    
                // Go down
                case 1:
                    if (row == m - layer - 1) {
                        dir = 2;
                        col--;
                    } else {
                        row++;
                    }
                    break;
                    
                // Go left
                case 2:
                    if (col == layer) {
                        dir = 3;
                        row--;
                    } else {
                        col--;
                    }
                    break;
                
                // Go up
                case 3:
                    if (row == layer + 1) {
                        dir = 0;
                        col++;
                        layer++;
                    } else {
                        row--;
                    }
                    break;
            }
            
            result.add(A.get(row).get(col));
        }
        
        return result;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
    }
}

File: problem_0071/solution.cpp
/*
Spiral Order Matrix I

Given a matrix of m * n elements (m rows, n columns), return all elements of the matrix in spiral order.

Example:

Given the following matrix:

[
    [ 1, 2, 3 ],
    [ 4, 5, 6 ],
    [ 7, 8, 9 ]
]

You should return

[1, 2, 3, 6, 9, 8, 7, 4, 5]
*/

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <cstdio>
#include <cstdlib>
#include <ctime>
#include <climits>
#include <cstring>
#include <functional>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <iterator>
#include <limits>
#include <list>
#include <map>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

#define ll long long
#define ull unsigned long long

using namespace std;

class Solution {
public:
    vector<int> spiralOrder(const vector<vector<int> > &matrix) {
        int rows = matrix.size();
        if (rows == 0) return vector<int> ();
        int cols = matrix[0].size();
        int row = 0, col = 0, layer = 0;
        vector<int> res;
        res.push_back(matrix[0][0]);
        int dir = 1;
        for (int step = 1; step < rows * cols; step++) {
            switch(dir) { // based on dir, check and change dir, then move on
            case 1: // supposed to go right
                if (col == cols - layer - 1) { // reach right bound
                    row++;
                    dir = 2;
                }
                else col++;
                break;
            case 2: // supposed to go down
                if (row == rows - layer - 1) { // reach downside bound
                    col--;
                    dir = 3;
                }
                else row++;
                break;
            case 3: // supposed to go left
                if (col == layer) { // reach left bound
                    row--;
                    dir = 4;
                }
                else col--;
                break;
            case 4: // supposed to go up
                if (row == layer + 1) { // reach upside bound
                    col++;
                    dir = 1;
                    layer++;
                }
                else row--;
                break;
            }
            res.push_back(matrix[row][col]);
        }
        return res;
    }
};

int main()
{
    return 0;
}

